Add documentation for FunctionConfig CRD and reconciler (#976)
nephio-project/porch: docs/content/en/docs/5_architecture_and_components/function-runner/interactions.md

Reviewer: As for task handler responsibilites, it rendes ppipeline orchestration, it's responsible for function sequencing, result aggregation, and RenderStatus creation.

Reviewer: It has many benefits. It is possible to use horizontal scaling, it is restart safe, which means no state loss, and it has a simple deployment model.

## Overview

> The Function Runner is a **separate gRPC service** that interacts with multiple systems: the Task Handler (via gRPC), Kubernetes API (for pod management and FunctionConfig CRDs), container registries (for image metadata), and wrapper servers (for function execution). It operates independently from the Porch server, enabling isolated function execution.

Reviewer 2: Is Task Handler concept still valid? Should we look to rename/redefine it?

Author: It "handles function evaluation" through grpc, but that isn't really a "system" it interacts with. I think more accurately, it would be "gRPC (for handling function evaluation requests)" or something.

Reviewer 2: Ye the mention of "Task" caught my attention really. Maybe just a re wording needed.

│                        ┌──────────────┐     ┌──────────┐     │
│                        │ Function Pods│     │ Registry │     │
│                        │  + Services  │     │   APIs   │     │
│                        └──────────────┘     └──────────┘     │
│                                                              │
└──────────────────────────────────────────────────────────────┘
```

External Resources:
- FunctionConfig CRDs (define executor configurations)
- ServiceTemplate CRDs (define pod/service templates)

## FunctionConfig Reconciler and Cache

The Function Runner includes an embedded reconciler that watches FunctionConfig CRDs and maintains an internal cache of executor configurations.

### Reconciliation Flow

```
FunctionConfig CRD
        ↓
  Watch Event
        ↓
  Reconciler
        ↓
  Parse Spec
        ↓
  ┌────┴────┬──────────┬──────────┐
  ↓         ↓          ↓          ↓
Pod     Binary      Go        Update
Config  Config    Config     Status
  ↓         ↓          ↓          ↓
Pod     Binary      Go        Return
Cache   Cache      Cache
  ↓         ↓          ↓
  └────┬────┴──────────┘
       ↓
  Ready for
  Evaluation
```

> **Reconciler responsibilities:**

Reviewer 2: Some repetition across the docs. Should this be defined once in a controller section? Or is it a controller? When does a Reconciler become a controller?

Author: Yepp, should probably just give it it's own page.

- Watch FunctionConfig resources across all namespaces
- Parse executor configurations (pod, binary, Go)
- Populate executor-specific caches
- Update FunctionConfig status with observed generation

> - Handle configuration errors and report via status

Reviewer: This reconciler has various responsibilites. It watches FunctionConfig resources across all namespaces and parses executor configurations (pod, binary, Go). The reconciler populates executor-specific caches, updates FunctionConfig status with observed generation, and handles configuration errors and reports via status.

**Cache structure:**
- **Pod executor cache**: Maps image names (with tags/prefixes) to `PodExecutorConfig` (TTL, parallelism, template overrides)
- **Binary executor cache**: Maps image names (with tags) to local binary paths
- **Go executor cache**: Maps image names (with tags) to in-process Go function processors

### Executor Selection Logic

When evaluating a function, the executor selection follows this pattern:

```
EvaluateFunction Request
        ↓
  Extract Image Name
        ↓
  Query Executor Cache
        ↓
  ┌────┴────┬──────────┬──────────┐
  ↓         ↓          ↓          ↓
Go      Binary     Pod       Not Found
Match   Match     Match       Error
  ↓         ↓          ↓          ↓
Go      Binary     Pod       Fallback
Exec    Exec      Exec      to Next
```

**Selection characteristics:**
- Image name extracted from request (including registry prefix and tag)
- Cache lookup considers full image reference with version
- Single function image can have multiple executors for different versions
  - Example: `apply-replacements:v0.1.0` → binary executor
  - Example: `apply-replacements:v0.2.0` → pod executor
- First matching executor is used

> - NotFoundError triggers fallback to next evaluator in chain

Reviewer: The image name is extracted from request (including registry prefix and tag), and the cache lookup considers full image reference with version. A single function image can have multiple executors for different versions:
  - Example: `apply-replacements:v0.1.0` → binary executor
  - Example: `apply-replacements:v0.2.0` → pod executor

The first matching executor is used and NotFoundError triggers fallback to the next evaluator in the chain.

- Wrapper server executes function binary and returns results

> - Multiple evaluations can execute in parallel on the same pod

Reviewer: Pod executor configuration comes from FunctionConfig resources and the pod cache is checked for existing pod (reuse if available). The pod selection uses round-robin among pods with minimum waitlist length. Cache miss triggers pod creation with wrapper server and the ClusterIP service provides stable DNS-based access. gRPC is connected to wrapper server in pod, which executed the function binary and returns the results. Multiple evaluations can be executed in parallel on the same pod.

**Configuration from FunctionConfig:**

> - `podExecutor.timeToLive`: How long pods remain cached before garbage collection

Reviewer: ```suggestion
- `podExecutor.timeToLive`: The duration of pods remaining cached before garbage collection
```

```

**Execution pattern:**
- Binary executor configuration comes from FunctionConfig resources
- Cache maps image names (with tags) to local binary paths
- Fast O(1) lookup by image name
- Direct process execution with ResourceList input

> - NotFoundError triggers fallback in multi-evaluator

Reviewer: The binary executor configuration comes from FunctionConfig resources. The cache maps image names (with tags) to local binary paths and fast O(1) lookup is performed by image name. This execution is a direct process execution with ResourceList input. NotFoundError triggers fallback in multi-evaluator.

**Configuration from FunctionConfig:**
- `binaryExecutor.tags[]`: Image tags mapped to this binary
- `binaryExecutor.path`: Binary path (absolute or relative to `--functions` dir)
- Multiple tags can map to same binary (e.g., `v0.1.0`, `v0.1`, specific digest)

**For detailed function evaluation, see [Function Evaluation]({{% relref "/docs/5_architecture_and_components/function-runner/functionality/function-evaluation.md" %}}).**

### Go-Based Execution

```
gRPC Request
        ↓
  Go Evaluator (in porch-server)
        ↓
  Lookup in Cache
        ↓
  Found? ──No──> Return NotFoundError
        │
       Yes
        ↓
  Call Go Function Processor
        ↓
  Return Response
```

**Execution pattern:**
- Go executor provides highest performance by avoiding process/container overhead
- Functions executed as native Go function calls within porch-server process
- Cache maps image names to `ResourceListProcessor` implementations
- Direct in-process execution with ResourceList input

> - NotFoundError triggers fallback to next evaluator (typically gRPC runtime)

Reviewer: The Go executor provides highest performance by avoiding process/container overhead. The functions are executed as native Go function calls within porch-server process. The cache maps image names to `ResourceListProcessor` implementations. This execution is a direct in-process execution with ResourceList input. NotFoundError triggers fallback to next evaluator (typically gRPC runtime).

**Configuration from FunctionConfig:**
- `goExecutor.tags[]`: Image tags mapped to this Go function
- `goExecutor.id`: Internal registration ID (defaults to `spec.image`)
- Registered functions: `apply-replacements`, `set-namespace`, `starlark`

**Benefits:**
- No container startup overhead
- No gRPC network calls
- Fastest possible execution for compatible functions

> - Lower resource usage compared to pod-based execution

Reviewer: This execution has various benefits. It has no container startup overhead or gRPC network calls. It is the fastest possible execution for compatible functions and has lower resource usage compared to pod-based execution.

        ↓
  ┌──────┴──────┬──────────┬─────────┐
  ↓             ↓          ↓         ↓
Pod Ops    Service Ops    CRs     Secrets
  ↓             ↓          ↓         ↓
Create/Get  Create/Get   Function   Auth
Delete      Delete       Config     Config
```

**Resource operations:**
- **Pods**: Create from template, get status, list by label, delete
- **Services**: Create ClusterIP frontend, get endpoints, delete
- **FunctionConfig CRs**: Watch for configuration changes
- **ServiceTemplate CRs**: Base pod and service templates
- **PodTemplate resources**: Base pod specifications
- **Secrets**: Access registry authentication and TLS certificates

**Template system:**
- ServiceTemplate CRs define base pod and service templates
- PodTemplate resources provide base pod specifications
- Inline templates as fallback if CRs/resources not found
- FunctionConfig provides per-function template overrides

> - Template version tracked for pod replacement on changes

Reviewer: The ServiceTemplate CRs define base pod and service templates and the PodTemplate resources provide base pod specifications. The template system uses inline templates as fallback if CRs/resources not found. Per-function template overrides is provided by FunctionConfig. Template version is tracked for pod replacement on changes.

- Pod and service lifecycle management
- Image metadata caching
- Registry authentication

> - FunctionConfig reconciliation and caching

Reviewer: The function runner has many responsibilites. It maintains (?) a gRPC service for function execution, performs evaluator selection and orchestration, and pod and service lifecycle management. It also perform image metadata caching, registry authentication, and FunctionConfig reconciliation and caching.

- No persistent state between requests
- Pod cache is ephemeral (TTL-based)
- Image cache is in-memory only
- Executor configuration cache populated from FunctionConfig CRs

> - Evaluators are stateless

Reviewer: When it comes to stateless service, there is no persistent state between requests. The pod cache is ephemeral (TTL-based) and the image cache is in-memory only. The executor configuration cache is populated from FunctionConfig CRs. Additionally, the evaluators are stateless.

- Garbage collection runs periodically
- TTL updates are asynchronous patches
- Cache warming at startup is concurrent

> - FunctionConfig reconciliation is event-driven

Reviewer: Tod creation happens in background goroutines while garbage collection runs periodically. The TTL updates are asynchronous patches. The cache warming at startup is concurrent and FunctionConfig reconciliation is event-driven.
